File: 0826-most-profit-assigning-work/0826-most-profit-assigning-work.py
class Solution:
    def maxProfitAssignment(self, diff: List[int], prof: List[int], wrk: List[int]) -> int:
        zipped = sorted(zip(diff, prof))
        for i in range(1, len(diff)):
            zipped[i] = (zipped[i][0], max(zipped[i][1], zipped[i-1][1]))
        
        ans, p = 0, [i[0] for i in zipped]
        
        for w in wrk:
            ind = bisect_right(p, w)
            if ind:
                ans += zipped[ind-1][1]
        return ans

    
# class Solution:
#     def maxProfitAssignment(self, diff: List[int], prof: List[int], wrk: List[int]) -> int:
        
#         q = sorted([(diff[ind], ind) for ind in range(len(diff))])
#         prof = [prof[j] for i,j in q]
#         for i in range(1, len(prof)):
#             prof[i] = max(prof[i-1], prof[i])
        
#         p, ans = [i[0] for i in q], 0
#         for w in wrk:
#             ind = bisect_right(p, w)
#             if ind!=0:
#                 ans += prof[ind-1]
#         return ans

File: 0826-most-profit-assigning-work/0826-most-profit-assigning-work.java
class Solution {
    public int maxProfitAssignment(int[] diff, int[] pr, int[] wrk) {
        int n = diff.length;
        
        int[][] pair = new int[n][2];
        for(int i=0; i<n; ++i) pair[i] = new int[]{diff[i], pr[i]};
        Arrays.sort(pair, (a, b) -> a[0]==b[0] ? a[1]-b[1] : a[0]-b[0]);
        
        int[] mxpr = new int[n+1];
        for(int i=1; i<=n; ++i) mxpr[i] = Math.max(mxpr[i-1], pair[i-1][1]);
        
        int ans = 0;
        for(int i=0; i<wrk.length; ++i){
            int idx = bs(pair, wrk[i]);
            ans += mxpr[idx+1];
        }
        
        return ans;
    }
    
    int bs(int[][] pair, int x){
        int lo = 0;
        int hi = pair.length-1;
        while(lo <= hi){
            int mid = lo + (hi-lo)/2;
            if(pair[mid][0] <= x) lo = mid + 1;
            else hi = mid -1;
        }
        return hi;
    }
}